File: Day4/soln4.py
from dataclasses import dataclass

inFile = open('../inputs/Day4.txt')
lines = inFile.readlines()

class Square:
  def __init__(self, num, state, row):
    self.num = num
    self.state = state
    self.row = row

draws = lines[0].strip().split(',')
for d in range(0, len(draws)):
    draws[d] = int(draws[d])

boards = []

def checkRow(board, row):
    for s in board[row]:
        if s.state == False:
            return False
    return True

def checkCol(board, col):
    for r in board:
        if r[col].state == False:
            return False
    return True
    

del lines[0]
del lines[0]
newBoard = []
for l in lines:
    if(l == '\n'):
        boards.append(newBoard)
        newBoard = []
        continue

    row = l.strip().split()
    squares = []
    for r in range(0, len(row)):
        squares.append(Square(int(row[r]), False, len(newBoard)))
    newBoard.append(squares)

boards.append(newBoard)

def sumRow(row):
    total = 0
    for s in row:
        if s.state == False:
            total += s.num
    return total

def sumCol(board, col):
    total = 0
    for r in baord:
        if r[col].state == False:
            total += r[col].num
    return total

def sumBoard(board):
    total = 0
    for r in board:
        total += sumRow(r)
    return total

def helper(draws, boards):
    for d in draws:
        for b in boards:
            for row in b:
                for r in range(0, len(row)):
                    if row[r].num == d:
                        row[r].state = True
                        if checkRow(b, row[r].row):
                            print("First: " + str(sumBoard(b) * d))
                            return
                        if checkCol(b, r):
                            print("First: " + str(sumBoard(b) * d))
                            return

helper(draws, boards)

def helper2(draws, boards):
    for d in draws:
        length = len(boards)
        deleted = 0
        for b in range(0, length):
            for r in boards[b - deleted]:
                for s in range(0, len(r)):
                    if r[s].num == d:
                        r[s].state = True
                        if checkRow(boards[b - deleted], r[s].row) or checkCol(boards[b - deleted], s):
                            if len(boards) == 1:
                                print("Second: " + str(sumBoard(boards[0]) * d))
                                return
                            del boards[b - deleted]
                            deleted += 1

# Reset for part 2
for b in boards:
    for r in b:
        for s in range(0, len(r)):
            r[s].state = False

helper2(draws, boards)

File: Day4/soln4.c
#include <stdio.h>
#include "fileUtils.h"
#include "stringUtils.h"
#include <string.h>
#include <math.h>
#include <stdlib.h>
#include <stdbool.h>

#define BOARD_SIDE_LENGTH 5

struct Square {
    int val;
    bool marked;
};
typedef struct Square Square;

struct Board {
    Square **grid;
    bool won;
};
typedef struct Board Board;

Square * genRow(char * row) {
    int length;
    int * vals = stringToIntArr(row, " ", &length);
    Square * squares = (Square *) malloc(length * sizeof(Square));

    for(int i = 0; i < length; i++) {
        squares[i].val = vals[i];
        squares[i].marked = false;
    }

    return squares;
}

bool checkRow(Square * row) {
    for(int i = 0; i < BOARD_SIDE_LENGTH; i++) {
        if(!row[i].marked) {
            return false;
        }
    }
    return true;
}

bool checkCol(Square ** grid, int col) {
    for(int i = 0; i < BOARD_SIDE_LENGTH; i++) {
        if(!grid[i][col].marked) {
            return false;
        }
    }
    return true;
}

bool won(Board * b) {
    for(int i = 0; i < BOARD_SIDE_LENGTH; i++) {
        if(checkRow(b->grid[i]) || checkCol(b->grid, i)) {
            b->won = true;
            return true;
        }
    }
    return false;
}

void mark(Board * b, int val) {
    for(int x = 0; x < BOARD_SIDE_LENGTH; x++) {
        for(int y = 0; y < BOARD_SIDE_LENGTH; y++) {
            if(b->grid[x][y].val == val) {
                b->grid[x][y].marked = true;
            }
        }
    }
}

int sumBoard(const Board b) {
    int total = 0;
    for(int x = 0; x < BOARD_SIDE_LENGTH; x++) {
        for(int y = 0; y < BOARD_SIDE_LENGTH; y++) {
            if(!b.grid[x][y].marked) {
                total += b.grid[x][y].val;
            }
        }
    }
    return total;
}

bool lastBoard(const Board * boards, int length) {
    for(int i = 0; i < length; i++) {
        if(!boards[i].won) {
            return false;
        }
    }
    return true;
}

void playBingo(Board * boards, int * draws, int boardNum, int length) {
    bool first = true;
    for(int i = 0; i < length; i++) {
        for(int b = 0; b < boardNum; b++) {
            if(boards[b].won) {
                continue;
            }

            mark(&boards[b], draws[i]);

            if(won(&boards[b])) {
                if(first) {
                    int sum = sumBoard(boards[b]);
                    printf("First: %d\n", sum * draws[i]);
                    first = false;
                }
                if(lastBoard(boards, boardNum)) {
                    printf("Second: %d\n", sumBoard(boards[b]) * draws[i]);
                    return;
                }
            }
        }
    }
}

int main() {
    int length;
    char ** lines = getInputLines(4, &length);
    int fileLength = length;
    int boardNum = ceil((float)(length - 2)/6);

    int *draws = stringToIntArr(lines[0], ",", &length);

    char *newline = lines[1];

    Board *boards = (Board *) malloc(boardNum * sizeof(Board));
    for(int b = 0; b < boardNum; b++) {
        boards[b].grid = malloc(BOARD_SIDE_LENGTH * sizeof(Square *));
        boards[b].won = false;
        for(int i = 0; i < BOARD_SIDE_LENGTH; i++) {
            boards[b].grid[i] = malloc(BOARD_SIDE_LENGTH * sizeof(Square));
        }
    }

    int boardIndex = 0;
    int rowIndex = 0;
    for(int i = 2; i < fileLength; i++) {
        if(strcmp(lines[i], newline) == 0) {
            rowIndex = 0;
            boardIndex++;
        }
        else {
            boards[boardIndex].grid[rowIndex] = genRow(lines[i]);
            rowIndex++;
        }
    }

    playBingo(boards, draws, boardNum, length);
    return 0;
}
